Add test to check kubernetes version

kasturinarra opened the pull request:
Adding this test to validate kubernetes version so that we do not ship any GA bits which include invalid version of kubernetes.

kasturinarra commented:
@zhouying7780 could you please help review ? Below link contains the log to the case.

https://privatebin-it-iso.int.open.paas.redhat.com/?5e5615f9207410d9#6Tj7wJ3KvTLckNe3YQEDwHvuu7BE9XUtLfySaJyYHr5U

zhouying7780 commented:
/lgtm

kasturinarra commented:
@soltysh can you please review the PR ? Thanks !!

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		version, err := ocAdmin.Run("version").Output()
+		o.Expect(version).To(o.ContainSubstring(kversionNum))
+		if err != nil {
+			e2e.Failf("Fail to get the Kubernetes version")
```
Make this error more descriptive.

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		}
+		re := regexp.MustCompile(`Kubernetes Version: v[0-9]+\.[0-9]+\.[0-9]+($|\+)`)
+		if re == nil {
+			e2e.Failf("version regexp err!")
```
Same here.

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		node := strings.Fields(nodeName)
+
+		for _, v := range node {
+			kubeletVersion, err := ocAdmin.WithoutNamespace().Run("get").Args("node", fmt.Sprintf("%s", v), "-o", "yaml", "-o", "custom-columns=VERSION:.status.nodeInfo.kubeletVersion").Output()
```
You can replace this with a single `oc get no -ocustom-columns=VERSION:.status.nodeInfo.kubeletVersion` and then just iterate over result ignoring the first row.

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		}
+		result := re.FindAllStringSubmatch(version, -1)
+		if match, _ := regexp.MatchString(kversionNum, result[0][0]); !match {
+			e2e.Failf("Failed to match Kuberntes version %v & OC Version %v", kversionNum, result[0][0])
```
nit: lowercase oc

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+	oc := exutil.NewCLI("oc-version")
+	ocAdmin := oc.AsAdmin()
+
+	g.It("check kubernetes version", func() {
```
```suggestion
	g.It("check kubernetes version matches the io.openshift.build.versions label ", func() {
```

kasturinarra replied:
Done.

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		o.Expect(err).NotTo(o.HaveOccurred())
+		podList := strings.Split(podNames, " ")
+
+		imageName, err := ocAdmin.Run("get").Args("pod", "-n", "openshift-kube-controller-manager", podList[0], "-o=jsonpath='{.spec.containers[0].image}'").Output()
```
Squash above two invocations into a single one:
`oc get po -n openshift-kube-controller-manager -lapp=kube-controller-manager -o=jsonpath='{..spec.containers[0].image}'` and grab the first output as the image name

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		imageName, err := ocAdmin.Run("get").Args("pod", "-n", "openshift-kube-controller-manager", podList[0], "-o=jsonpath='{.spec.containers[0].image}'").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		masterNode, err := exutil.GetFirstMasterNode(oc)
```
Drop the helper and replace that with:
```
masterNodes, err := oc.AdminKubeClient().CoreV1().Nodes().List(context.Background(), metav1.ListOptions{
		LabelSelector: `node-role.kubernetes.io/master`,
})
o.Expect(err).NotTo(o.HaveOccurred())
```
grab the first element in the list.

soltysh replied:
Or do similar get as you do with `oc get nodes`.

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,76 @@
+		masterNode, err := exutil.GetFirstMasterNode(oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		imageInfo, err := exutil.DebugNodeWithChroot(oc, masterNode, "oc", "image", "info", "--registry-config=/var/lib/kubelet/config.json", strings.Trim(imageName, "'"))
```
Drop the helper and just do inline:
```
out, err := oc.AsAdmin().Run("debug").Args("node/"+masterName, "--", "chroot", "/host", "oc", "image", "info", "--registry-config=/var/lib/kubelet/config.json", strings.Trim(imageName, "'")).Output()
```

kasturinarra replied:
Done

soltysh commented on test/extended/util/nodes.go:
```diff
@@ -0,0 +1,37 @@
+package util
```
Those helpers should not be needed.

kasturinarra replied:
Done

kasturinarra commented:
@soltysh addressed all comments, could you please help review & merge ? Thanks !!

Logs: https://privatebin-it-iso.int.open.paas.redhat.com/?01acac39e3194b4f#418RrqAea8Ax78F4SPwx3472vasm5aJUug78UVwmkBsz

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,71 @@
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	"strings"
```
Nit: move builtin imports together above all others.

kasturinarra replied:
My bad !! Thanks !!

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,71 @@
+		masterName := strings.Fields(masterNode)
+
+		imageInfo, err := oc.AsAdmin().Run("debug").Args("node/"+masterName[0], "--", "chroot", "/host", "oc", "image", "info", "--registry-config=/var/lib/kubelet/config.json", strings.Fields(imageName)[0]).Output()
+
```
Nit: not needed empty space

kasturinarra replied:
Done. Thanks !!

soltysh reviewed (approved):
/lgtm
/approve
/retest-required

soltysh reviewed (approved):
/lgtm

soltysh commented:
/hold
it looks like there's a problem with this PR

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,69 @@
+		re1 := regexp.MustCompile(`io.openshift.build.versions=kubernetes=[1-9]{1}.[1-9]{1,2}.[1-9]{1}`)
+		kversion := re1.FindAllString(imageInfo, -1)
+		kversion = strings.Split(kversion[0], "=")
+		kversionNum := kversion[len(kversion)-1]
```
@kasturinarra it looks like you'll need to add checks here to ensure that `kversion` has appropriate length and similarly below when you're accessing `result`.

soltysh commented:
/test e2e-gcp

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+			if err != nil {
+				e2e.Failf("Fail to get the kubelet version")
+			}
+			re := regexp.MustCompile(`v[0-9]+\.[0-9]+\.[0-9]+($|\+)`)
```
Same here: `[1-9]{1}.[1-9]{1,2}.[1-9]{1,2}`

kasturinarra replied:
Done

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+			if re == nil {
+				e2e.Failf("kubeletVersion regexp err!")
+			}
+			kubeletResult := re.FindAllStringSubmatch(value, -1)
```
Again `FindAllString`

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+				e2e.Failf("kubeletVersion regexp err!")
+			}
+			kubeletResult := re.FindAllStringSubmatch(value, -1)
+			if match, _ := regexp.MatchString(kversionNum, kubeletResult[0][0]); !match {
```
`kubeletResult[0]`

soltysh commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+				e2e.Failf("kubeletVersion regexp err!")
+			}
+			kubeletResult := re.FindAllStringSubmatch(value, -1)
+			if match, _ := regexp.MatchString(kversionNum, kubeletResult[0][0]); !match {
```
Also you're missing check if len(kubeletResult) == 0 {... // fail`

kasturinarra replied:
Done, added

kasturinarra commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		re1 := regexp.MustCompile(`io.openshift.build.versions=kubernetes=[1-9]{1}.[1-9]{1,2}.[1-9]{1}`)
```
Done

kasturinarra commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+		if err != nil {
+			e2e.Failf("Failed to match output from oc version %v to io.openshift.build.versions label %v", version, kversionNum)
+		}
+		re := regexp.MustCompile(`Kubernetes Version: v[0-9]+\.[0-9]+\.[0-9]+($|\+)`)
```
Done

kasturinarra commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+		if re == nil {
+			e2e.Failf("Failed to match regex, please check kubernetes version to see if it is one release candidate")
+		}
+		result := re.FindAllStringSubmatch(version, -1)
```
Done

kasturinarra commented on test/extended/cli/version.go:
```diff
@@ -0,0 +1,74 @@
+		if len(result) == 0 {
+			e2e.Failf("Failed to find match for kubernetes version in result %s", result)
+		}
+		if match, _ := regexp.MatchString(kversionNum, result[0][0]); !match {
```
Done

soltysh reviewed (approved):
/hold cancel
/lgtm
/approve
